File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def bfs(start, end, maps):
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    n, m = len(maps), len(maps[0])
    
    visited = [[False] * len(maps[0]) for _ in maps]
    
    q = deque()
    q.append((start[0], start[1], 1))  # 시작 좌표와 거리(1)
    
    while q:
        x, y, dist = q.popleft()
        
        if (x, y) == end:
            return dist
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]    
            
            if 0 <= nx < n and 0 <= ny < m:
                if not visited[nx][ny] and maps[nx][ny] == 1:
                    q.append((nx, ny, dist + 1))
                    visited[nx][ny] = True
    
    return -1  # 도착 지점에 도달할 수 없는 경우

def solution(maps):
    maps_n = len(maps)
    maps_m = len(maps[0])
    
    answer = bfs((0, 0), (maps_n - 1, maps_m - 1), maps)
    
    return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};
    int n, m;
    
    public int solution(int[][] maps) {
        n = maps.length;
        m = maps[0].length;
        
        int answer = bfs(0, 0, maps);
        
        return answer;
    }
    
    private int bfs(int startX, int startY, int[][] maps) {
        Deque<int[]> q = new ArrayDeque<>();
        q.add(new int[]{startX, startY});
        
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];
            
            if (x == n - 1 && y == m - 1) return maps[x][y];
            
            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];
                
                if (nx < 0 || ny < 0 || nx >= n || ny >= m) continue;
                if (maps[nx][ny] == 1) {
                    maps[nx][ny] += maps[x][y];
                    q.add(new int[]{nx, ny});   
                }
            }
        }
        
        return -1;
    }
}